### PR TITLE
fix for premature childShard lease creation

### PR DESCRIPTION
**Pull request by ychunxue**
*Issue #, if available:*

*Description of changes:*

Fix for premature childShard lease creation.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ShutdownTask.java [hunk 1]
```diff
@@ -191,7 +159,67 @@ public TaskResult call() {
...
+                if (currentLease == null) {
+                    throw new InvalidStateException(shardInfo.getShardId()
+                                                            + " : Lease not owned by the current worker. Leaving ShardEnd handling to new owner.");
+                }
+                if (!CollectionUtils.isNullOrEmpty(childShards)) {
+                    createLeasesForChildShardsIfNotExist();
+                    updateCurrentLeaseWithChildShards(currentLease);
+                    recordProcessorCheckpointer.setSequenceNumberAtShardEnd(
+                            recordProcessorCheckpointer.getLargestPermittedCheckpointValue());
+                    recordProcessorCheckpointer.setLargestPermittedCheckpointValue(ExtendedSequenceNumber.SHARD_END);
```
Comment (ashwing) [inline src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ShutdownTask.java]: These should be outside of this if block. Wait for Josh's lease cleanup PR. We need to rebase on top of that.

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ShutdownTask.java [hunk 2]
```diff
@@ -44,6 +47,8 @@
     private static final Log LOG = LogFactory.getLog(ShutdownTask.class);
 
     private static final String RECORD_PROCESSOR_SHUTDOWN_METRIC = "RecordProcessor.shutdown";
+    @VisibleForTesting
+    static final int RETRY_RANDOM_MAX_RANGE = 10;
```
Comment (ashwing) [inline src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ShutdownTask.java]: Let's change this to 50

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ShutdownTask.java [hunk 3]
```diff
@@ -106,113 +117,171 @@
      */
     @Override
     public TaskResult call() {
+        MetricsHelper.startScope(metricsFactory, SHUTDOWN_TASK_OPERATION);
```
Comment (ashwing) [inline src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ShutdownTask.java]: Let's not introduce new metrics as part of this PR.

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ShutdownTask.java (hunk 2 above)
Comment (ashwing) [reply src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ShutdownTask.java]: Let's fix this

File: src/main/java/com/amazonaws/services/kinesis/leases/impl/LeaseSerializer.java [hunk 4]
```diff
@@ -137,6 +137,18 @@ public Lease fromDynamoRecord(Map<String, AttributeValue> dynamoRecord) {
         return result;
     }
 
+    @Override
+    public Map<String, ExpectedAttributeValue> getDynamoExistantExpectation(final String leaseKey) {
+        Map<String, ExpectedAttributeValue> result = new HashMap<>();
+
+        ExpectedAttributeValue expectedAV = new ExpectedAttributeValue();
+        expectedAV.setValue(DynamoUtils.createAttributeValue(leaseKey));
+        expectedAV.setExists(true);
```
Comment (ashwing) [inline src/main/java/com/amazonaws/services/kinesis/leases/impl/LeaseSerializer.java:146]: If builder is available let's use it

https://github.com/ashwing/amazon-kinesis-client/pull/88/files#diff-4705950dfc6a2c55e5d2677225cf531fR199

File: src/main/java/com/amazonaws/services/kinesis/leases/interfaces/ILeaseSerializer.java [hunk 5]
```diff
@@ -79,6 +79,13 @@
      */
     public Map<String, ExpectedAttributeValue> getDynamoNonexistantExpectation();
 
+    /**
+     * @return the attribute value map asserting that a lease does exist.
+     */
+    default Map<String, ExpectedAttributeValue> getDynamoExistantExpectation(final String leaseKey) {
```
Comment (ashwing) [inline src/main/java/com/amazonaws/services/kinesis/leases/interfaces/ILeaseSerializer.java]: nit : existent

File: src/main/java/com/amazonaws/services/kinesis/leases/impl/LeaseManager.java [hunk 6]
```diff
@@ -620,7 +620,7 @@ public void updateLeaseWithMetaInfo(T lease, UpdateField updateField)
         UpdateItemRequest request = new UpdateItemRequest();
         request.setTableName(table);
         request.setKey(serializer.getDynamoHashKey(lease));
-        request.setExpected(serializer.getDynamoLeaseCounterExpectation(lease));
+        request.setExpected(serializer.getDynamoExistantExpectation(lease.getLeaseKey()));
```
Comment (ashwing) [inline src/main/java/com/amazonaws/services/kinesis/leases/impl/LeaseManager.java]: Why are we replacing leasecounter expectation with leaseKey?

Comment (ashwing) [reply src/main/java/com/amazonaws/services/kinesis/leases/impl/LeaseManager.java]: Also can we add these two integ tests?

https://github.com/ashwing/amazon-kinesis-client/pull/88/files#diff-d664625d66f92ee877c89352871066abR110

**Review by ashwing — COMMENTED** [review]
Left comments

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ShutdownTask.java (hunk 2 above)
Comment (ychunxue) [reply src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ShutdownTask.java]: Addressed

File: src/main/java/com/amazonaws/services/kinesis/leases/impl/LeaseSerializer.java (hunk 4 above)
Comment (ychunxue) [reply src/main/java/com/amazonaws/services/kinesis/leases/impl/LeaseSerializer.java:146]: Builder is not available in this version of sdk

File: src/main/java/com/amazonaws/services/kinesis/leases/impl/LeaseManager.java (hunk 6 above)
Comment (ashwing) [reply src/main/java/com/amazonaws/services/kinesis/leases/impl/LeaseManager.java]: > Why are we replacing leasecounter expectation with leaseKey?

Synced up with Jessica offline. This change is fine.

File: src/main/java/com/amazonaws/services/kinesis/leases/interfaces/ILeaseSerializer.java (hunk 5 above)
Comment (ychunxue) [reply src/main/java/com/amazonaws/services/kinesis/leases/interfaces/ILeaseSerializer.java]: Addressed

**Review by ashwing — APPROVED** [review]
lgtm

**Pull request merged**